docs/hugo/content/design/ADR-2022-09-Multiple-Credential-Operator.md

```diff
@@ -0,0 +1,153 @@
+cache and perform the operation. If its a new resource, operator would have to go through the secret selection hierarchy. Where, operator would first check if Per-resource or per-resource-group secret exists, if not, then will check for 
+the namespaced secret. If any of the above is provided, operator would perform actions according to the **options** below. If none is provided, we'll fallback to use the global credential for that resource. 
+
+![hierarchy](images/adr-2022-09-multiple-credential-operator.png) 
+
+## Options
+
+In all the proposed solutions, we would need to cache the credentials to be used for resources when we reconcile and a mechanism to identify the credentials used for a resource. 
+
+### Option 1: Secrets lookup
+
+In secrets lookup option, we would need a fixed secret pattern like below. Which then will be looked up for each time a new resource is created. Operator would look up for the secret pattern for resource group first, then namespace. If secret exists, cache the credentials for the resource to be used for the next time or use the default global.
+
+NOTE: Annotation/condition will be added on the resource while creation which will have information about the secret name used for creation/reconcile.
+
+**Fixed secret name pattern:**
+```
+Namespaced secret -> aso-credential (in the namespace)
+Per-resource-group secret -> aso-credential-{rg-name}
+Per-resource secret -> aso-credential-{resource-name}
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly
+
+**Cons:**
+1. Not very flexible, as users don't have much control for which resource they want to use which credential
+
+### Option 2: Configuration using annotations
+
+In Configuration using annotations option, we'll use annotations like below on the resource to fetch credentials from a secret and cache it. Will default to global if secret with name not found or no annotation provided.
+
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Flexible, as users have the control for which resource they want to use which credential
+
+**Cons:**
+1. Will have to specify annotation on each resource explicitly
+
+### Option 3: Secret lookup + configuration using annotations
+
+Here, we would use a fixed pattern for namespaces secrets + annotations for per-resource-group and per-resource secrets. Operator would look up for the annotation(`serviceoperator.azure.com/credential-from`) on resource created, if annotation found, then use credentials from secret in annotation. If not, then look for the namespaced secret and use the namespaced credential.If none provided, operator shall continue with the global credential.
+
+NOTE: If namespaced secret or global secret is used, operator would inject `serviceoperator.azure.com/credential-from` annotation with the secret name value. 
```

Comment:
Does it make sense to inject this for the global secret? It's supposed to point to a secret in _this_ namespace, right? But the credential for the global secret is not in this namespace probably.  I think we shouldn't inject this for the global secret but should still inject the `serviceoperator.azure.com/subscription` annotation I mention above.

For the namespace scoped secret, we could inject this annotation. can you call out the win of doing so here though? I'm assuming it's so that the user understands which secret we've actually used with this resource (rather than it being some implicit thing where they need to go "oh does it not have an annotation, then let me check the namespace secret). Which I think I like, but if you have other reasons call them out here too.

Reply:
Would help users to know which secret is being used for a resource, and for the operator in a manner that we don't have to do the secret lookup each time for namespaced secret and know from where to use the credential on a reconcile.

Reply:
This brings up an interesting scenario - if what happens if a resource is created using a lower priority secret (e.g. a namespaced secret), but then the user creates a higher priority one (e.g. a resource group secret), should the operator continue to use the original one, or should it switch?

I can see migration scenarios where users have things set up one way and then create new secrets, expecting the operator to switch over. Does having annotations on the resource prevent this from working? How would the user do the migration?

Reply:
I think it would make sense to switch the secret used at that point @theunrepentantgeek. As you say, there are scenarios where they do this on purpose.

I do see some possible awkwardness with the annotations though, it would be good exploring these options more in depth in the design to see if we can come up with solutions for them.

1. Moving from global to per-namespace, there was no annotation to begin with so no issues there.
2. Moving from per-namespace to per-resourceGroup: There would be an automatically added annotation pointing to the per-namespace secret on all of the existing resources. If the user then changes the annotation on the ResourceGroup to point to a different secret, that does not change the an annotation on any of the other resources in that resourceGroup, so we have an awkward situation where RG points at secret A, but all other resources point at secret B. Who wins?

We could possibly solve that by, next time the child resources are reconciled, fixing the annotation to point to the same secret as the resourceGroup, but that seems awkward because now things aren't so simple as "the secret used will always be what is set in `serviceoperator.azure.com/credential-from`". Another problem with reusing the `serviceoperator.azure.com/credential-from` for us setting it and the user also having the option to set it, is that we need to watch that annotation to trigger reconciles, but if we can also set it ourselves in a reconcile it's possible that a reconcile runs, changes that value and then triggers another reconcile based off of that. So I've tried to keep annotations in a spec/status style, where it's either something we write and the user _never_ writes, or it's something the user writes and we _never_ write. We can even write those as long as it's in a webhook, (won't trigger extra reconciles), but I don't see a way to deal with this case really cleanly off the top of my head.

Is it clearer to _not_ annotate any resources other than ResourceGroups with the `serviceoperator.azure.com/credential-from` annotation automatically? So basically in our initial implementation that annotation would never be on anything other than `ResourceGroup` (and possibly other top-level resources we designate)?

Reply:
I think we SHOULD not suggest shifting credentials? Do we want to allow users to change the credential setting once the resource is created?

Should suggest re-creating the resource with new credentials rather. If user tries to change credential on already present resource, we should set condition to false.

Reply:
> I think we SHOULD not suggest shifting credentials?

Sorry, I don't understand this phrasing.
Are you saying we shouldn't comment on shifting credentials (our docs would be silent on the subject) or that we should document that shifting credentials is unsupported?

Reply:
> This brings up an interesting scenario - if what happens if a resource is created using a lower priority secret (e.g. a namespaced secret), but then the user creates a higher priority one (e.g. a resource group secret), should the operator continue to use the original one, or should it switch?

As on the reply to this, Operator would use whatever is present on annotation. The idea of adding annotation is that, we're not storing any state of which credential is being used for which resource. If annotation says `credential-from: secretA` then operator on reconcile, would simple fetch the credential for secretA(if exist) or run into failure if does not exist. The idea of adding annotation for namespaced secret is to eliminate the need of repetitive lookup and storing state.

Reply:
> > I think we SHOULD not suggest shifting credentials?
> 
> Sorry, I don't understand this phrasing. Are you saying we shouldn't comment on shifting credentials (our docs would be silent on the subject) or that we should document that shifting credentials is unsupported?

Sorry for being unclear there, we should document that it's not suggested, as it can lead to failures on reconcile. As the new credentials might not have permissions to update the resource.

Reply:
> Sorry for being unclear there, we should document that it's not suggested, as it can lead to failures on reconcile. As the new credentials might not have permissions to update the resource.

I think we should consider the "I want to change credentials used to reconcile this object" case. This is something that you can change pretty easily if you're say, running your own code or using something like the `az cli`. It seems to me that we should support it. If users attempt to change to a credential that doesn't have permissions to do what they wanna do, then that's their issue. It seems weird to block or not recommend that because they might make a mistake though. 

Any issue they could run into with changing credentials they could also run into on first-use (the credential they try to use is wrong). So we're going to need to make sure we give a good error in those cases either way.